test/library/gui/dialogs.lua

```diff
@@ -0,0 +1,172 @@
+    expect.eq(0, mock_cb2.call_count)
+
+    expect.eq(before_scr, dfhack.gui.getCurViewscreen(true), "Selecting an item should return us to previous screen")
+end
+
+function test.ListBox_closes_on_select2()
+    local before_scr = dfhack.gui.getCurViewscreen(true)
+
+    local args = {}
+    local mock_cb = mock.func()
+    local mock_cb2 = mock.func()
+    args.on_select = mock_cb
+    args.on_select2 = mock_cb2
+    args.choices = {{
+        text = 'ListBox_closes_on_select2'
+    }}
+
+    local lb = dialogs.ListBox(args)
+    lb:show()
+    wait()
+    send_keys('SEC_SELECT')
+    
+    expect.eq(0, mock_cb.call_count)
+    expect.eq(1, mock_cb2.call_count)
+
+    expect.eq(before_scr, dfhack.gui.getCurViewscreen(true), "Selecting an item should return us to previous screen")
+end
+
+function test.ListBox_stays_open_with_multi_select()
+    local before_scr = dfhack.gui.getCurViewscreen(true)
+
+    local args = {}
+    local mock_cb = mock.func()
+    local mock_cb2 = mock.func()
+    args.on_select = mock_cb
+    args.on_select2 = mock_cb2
+    args.dismiss_on_select = false
+    args.choices = {{
+        text = 'ListBox_stays_open_with_multi_select'
+    }}
+
+    local lb = dialogs.ListBox(args)
+    lb:show()
+    local lb_scr = dfhack.gui.getCurViewscreen(true)
+    wait()
+    send_keys('SELECT')
+
+    expect.eq(lb_scr, dfhack.gui.getCurViewscreen(true), "Selecting an item should NOT close the ListBox")
+
+    send_keys('SEC_SELECT')
+    expect.eq(before_scr, dfhack.gui.getCurViewscreen(true), "With default dismiss_on_select2 it should return us to previous screen")
+    
+    expect.eq(1, mock_cb.call_count)
+    expect.eq(1, mock_cb2.call_count)
+end
+
+function test.ListBox_stays_open_with_multi_select2()
+    local before_scr = dfhack.gui.getCurViewscreen(true)
+
+    local args = {}
+    local mock_cb = mock.func()
+    local mock_cb2 = mock.func()
+    args.on_select = mock_cb
+    args.on_select2 = mock_cb2
+    args.dismiss_on_select2 = false
+    args.choices = {{
+        text = 'ListBox_stays_open_with_multi_select2'
+    }}
+
+    local lb = dialogs.ListBox(args)
+    lb:show()
+    local lb_scr = dfhack.gui.getCurViewscreen(true)
+    wait()
+    send_keys('SEC_SELECT')
+
+    expect.eq(lb_scr, dfhack.gui.getCurViewscreen(true), "Sec-selecting an item should NOT close the ListBox")
+
+    send_keys('SELECT')
+    expect.eq(before_scr, dfhack.gui.getCurViewscreen(true), "With default dismiss_on_select it should return us to previous screen")
+    
+    expect.eq(1, mock_cb.call_count)
+    expect.eq(1, mock_cb2.call_count)
+end
+
+function test.ListBox_with_multi_select()
+    local before_scr = dfhack.gui.getCurViewscreen(true)
+
+    local args = {}
+    local mock_cb = mock.func()
+    local mock_cb2 = mock.func()
+    args.on_select = mock_cb
+    args.on_select2 = mock_cb2
+    args.dismiss_on_select = false
+    args.dismiss_on_select2 = false
+    args.choices = {{
+        text = 'ListBox_with_multi_select'
```

Comment:
This PR gives us a low-level option for the behavior of the list, but I'm wondering now (seeing the tests does help me visualize the usage patterns) if we need some more functionality around it. This really isn't "multi-select" per se. It's just allowing multiple single `submit()` calls without forcing the user to re-open the ListBox, except with no visual changes to the dialog to indicate what is happening.

If the goal here is to provide a true multi-select interface, the visuals and behavior might need modification.  For example, it should somehow visually indicate that multiple items are selected (e.g. via a highlight that is distinguishable from the cursor or perhaps just an X character to the left of the list item). The user should also be told somehow that LEAVESCREEN is the only way out of the dialog if both `dismiss_on_select` options have been set to true.

What do you think? Do you already have some code that would use this that I could look at for context? Perhaps we can move some of that code into the library and encapsulate the behavior in a MultiSelectListBox subclass.

Reply:
I've added a test demonstrating how I use it. The actual code where I use it belongs into the dfhack/scripts repository, probably in the materials.lua -- I'll get it there some other day.

I can make a MultiSelectListBox out of this, but there are more issues related to that - like FilteredList.setFilter not working by default anymore (see #1912). So I thought as the first step I'd propose the minimal possible change to achieve what I need :)

>  The user should also be told that LEAVESCREEN is the only way out of the dialog if both dismiss_on_select options have been set to true.

I'm not sure about that: pressing ESC to leave a screen is the natural way of doing it and works everywhere in DF. Additional advertising would only add visual clutter.

Reply:
I think I'm ok with leaving out `Esc` - we're already leaving out `Enter` and `Shift-Enter` in all ListBox instances.